src/GovUk.Education.ExploreEducationStatistics.Data.Services/SubjectCsvMetaService.cs

```diff
@@ -0,0 +1,182 @@
+using GovUk.Education.ExploreEducationStatistics.Data.Services.Security.Extensions;
+using GovUk.Education.ExploreEducationStatistics.Data.Services.Utils;
+using GovUk.Education.ExploreEducationStatistics.Data.Services.ViewModels.Meta;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.Logging;
+
+namespace GovUk.Education.ExploreEducationStatistics.Data.Services;
+
+public class SubjectCsvMetaService : ISubjectCsvMetaService
+{
+    private readonly ILogger<SubjectCsvMetaService> _logger;
+    private readonly StatisticsDbContext _statisticsDbContext;
+    private readonly ContentDbContext _contentDbContext;
+    private readonly IUserService _userService;
+    private readonly IFilterItemRepository _filterItemRepository;
+    private readonly IReleaseFileBlobService _releaseFileBlobService;
+
+    public SubjectCsvMetaService(
+        ILogger<SubjectCsvMetaService> logger,
+        StatisticsDbContext statisticsDbContext,
+        ContentDbContext contentDbContext,
+        IUserService userService,
+        IFilterItemRepository filterItemRepository,
+        IReleaseFileBlobService releaseFileBlobService)
+    {
+        _logger = logger;
+        _statisticsDbContext = statisticsDbContext;
+        _contentDbContext = contentDbContext;
+        _userService = userService;
+        _filterItemRepository = filterItemRepository;
+        _releaseFileBlobService = releaseFileBlobService;
+    }
+
+    public async Task<Either<ActionResult, SubjectCsvMetaViewModel>> GetSubjectCsvMeta(
+        ReleaseSubject releaseSubject,
+        ObservationQueryContext query,
+        IList<Observation> observations,
+        CancellationToken cancellationToken = default)
+    {
+        return await _userService.CheckCanViewSubjectData(releaseSubject)
+            .OnSuccess(() => GetCsvStream(releaseSubject, cancellationToken))
+            .OnSuccess(
+                async csvStream =>
+                {
+                    var locations = GetLocations(observations);
+                    var filters = await GetFilters(observations);
+                    var indicators = await GetIndicators(query);
+                    var headers = await ListCsvHeaders(csvStream, filters, indicators, locations);
+
+                    return new SubjectCsvMetaViewModel
+                    {
+                        Filters = filters,
+                        Locations = locations,
+                        Indicators = indicators,
+                        Headers = headers
+                    };
+                }
+            );
+    }
+
+    private async Task<Stream?> GetCsvStream(
+        ReleaseSubject releaseSubject,
+        CancellationToken cancellationToken = default)
+    {
+        try
+        {
+            var releaseFile = await _contentDbContext.ReleaseFiles
+                .Include(rf => rf.File)
+                .SingleAsync(
+                    predicate: rf =>
```

Comment:
Are we needing to use named parameters here?  I think that in general we only use them when we have multiple parameters of the same type that could potentially get mixed up in error e.g. multiple Guid parameters?

Reply:
I don't think there's any hard and fast rules tbh. I like adding named parameters in some cases where the readability is improved and I considered this to be one of them.